Allow :base to be a map of inherited options for defsetting

crisptrutski opened the pull request:
Closes: https://github.com/metabase/metabase/issues/38250

### Description

This is an alternate to [overloading type as a base type](https://github.com/metabase/metabase/pull/37863).

I'm not sure which one I prefer. Using a different key means we don't have to inherit the type, and avoids overloading a field, but on the other hand I feel that conflating "type" with "inheritance" is more natural and familiar.

Another idea is to riff off [how kwarg functions can accept maps](https://clojure.org/news/2021/03/18/apis-serving-people-and-programs). That might take a bit more macro foo, so I tried this first.

crisptrutski committed:
Add :base option to bundle setting options via inheritance

Co-authored-by: Cal Herries <39073188+calherries@users.noreply.github.com>

improve base setting tests

calherries commented on src/metabase/models/setting.clj:
```diff
@@ -618,9 +618,16 @@
   (str (random-uuid)))
 
+;; This var-type allows the bundling of a number of attributes together. In some sense it is defining a subtype / mixin.
+(def ^{:doc "A random uuid value that should never change again"} uuid-nonce-type
```
```suggestion
(def ^{:doc "A random uuid value that should never change again"} uuid-nonce-base
```

calherries commented on src/metabase/models/setting.clj:
```diff
@@ -618,9 +618,16 @@
+(defn- random-uuid-str []
   (str (random-uuid)))
 
+;; This var-type allows the bundling of a number of attributes together. In some sense it is defining a subtype / mixin.
```
```suggestion
;; This base allows the bundling of a number of attributes together. In some sense it is defining a subtype / mixin.
```

calherries commented on src/metabase/models/setting.clj:
```diff
@@ -1079,6 +1086,15 @@
+(defn- inherit-base
+  "Recursively uses the values in `:base` as default values for `setting-options`"
+  [{:keys [base] :as setting-options}]
+  (let [without-base (dissoc setting-options :base)]
+    (if-not base
+      without-base
+      (do (assert (map? base))
+          (merge (inherit-base base) without-base)))))
```
I don't think this should recurse. It's simpler without it and we have no need for it.
```suggestion
  (let [without-base (dissoc setting-options :base)]
    (merge base without-base))
```
Also I feel the assertion is unnecessary. `merge` with other types should fail

crisptrutski replied:
> I don't think this should recurse. 

Transitivity is something I would expect from an inheritance mechanism. Since we've gone the route of keeping type as separate concern, and users can still merge their own maps before passing to the macro, I'll go ahead and remove this. Using merge does mean losing safeguards like making read-only "final" etc, but perhaps this is YAGNI.

calherries reviewed (changes requested):
This is still my preferred approach. Remember to add `:base` to the `defsetting` docstring and write a test with a real `defsetting` to check that the option is merged in that considers the property might have a default, e.g. try setting two different values of `:cache?` with `:base` and check the override works

crisptrutski commented:
@calherries I think I've addressed all your feedback. Don't forget to review https://github.com/metabase/metabase/pull/38128, as it's a dependency of this branch.

qnkhuat commented on src/metabase/models/setting.clj:
```diff
@@ -618,9 +618,16 @@
   (str (random-uuid)))
 
+;; This base allows the bundling of a number of attributes together. In some sense it is defining a subtype / mixin.
+(def ^{:doc "A random uuid value that should never change again"} uuid-nonce-base
```
nit, you can put the docstring after the var name, no need for using meta tag

crisptrutski replied:
Ah that's nice! In older versions of Clojure this did not work.

qnkhuat commented on test/metabase/models/setting_test.clj:
```diff
@@ -199,6 +199,30 @@
+(def ^:private base-options
+  {:setter   :none
+   :type     :csv
+   :default "base-default"})
```
```suggestion
   :default  "base-default"})
```

qnkhuat commented on test/metabase/models/setting_test.clj:
```diff
@@ -199,6 +199,30 @@
+
+(defsetting test-base-setting
+  "Setting to test the `:feature` property of settings. This only shows up in dev."
+    :visibility :internal
+    :default "my,default"
+    :base base-options)
```
```suggestion
  :visibility :internal
  :default "my,default"
  :base base-options)
```

qnkhuat commented on test/metabase/models/setting_test.clj:
```diff
@@ -199,6 +199,30 @@
+    :base base-options)
+
+(deftest ^:parallel defsetting-with-base-test
+  (testing "A setting which specifies some base options"
```
```suggestion
  (testing "A setting which specifies a base option"
```

crisptrutski replied:
I prefer the original wording here, it is very intentional. The way I use the terms:

- Each key value pair is one option
- The base map is therefore a collection of options, or "base options"
- We are specifying "some" base options, as it is optional.

qnkhuat replied:
makes sense to me

qnkhuat commented on test/metabase/models/setting_test.clj:
```diff
@@ -199,6 +199,30 @@
+      (testing "Uses non-overridden base options"
+        (is (= :csv (:type setting)))
+        (is (= :none (:setter setting)))
+        ;; Obfuscate test-base-setting! to avoid a false positive from `deftest-check-parallel`
```
a better way to do this is to tell kondo to ignore it.

```
#_{:clj-kondo/ignore [linter-name]}
```

crisptrutski replied:
I think this only works when the linter reports a finding, but in this case the hook throws an exception. I'll give it a try though.

qnkhuat replied:
what's the key of that error? I'll see if I could fix it, looks like we need to update the hook to respect kondo/ignore

crisptrutski replied:
Ah I took another look and see that `error!` is just an alias for `reg-finding!`, which encapsulates the key `:metabase/validate-deftest`.

I'd thought Cursive would have shown me the key if there was one, and that `error!` wasn't sending one. Cool, I've pushed a version with the ignore. Unfortunately it seems that I have to ignore the entire test, which isn't ideal. I might actually prefer the symbol-builder approach if we need to extend this test in future.

crisptrutski replied:
Score one for Cider! I see that shows the ignore key front and center 😄

qnkhuat commented on src/metabase/models/setting.clj:
```diff
@@ -1079,6 +1086,14 @@
        (not= (:setter setting-definition) :none)
        (not (ns-in-test? (:namespace setting-definition)))))
 
+(defn- inherit-base
```
```suggestion
(defn- maybe-inherit-base
```

crisptrutski replied:
When I see a function starting with "maybe" I assume this means it returns something optional, but in this case it always returns a map. That said, the "inherit" verb makes it clear that this is a transformation, so I guess it's clear that it means it is "maybe" a no-op in this case. I don't think the extra verbiage adds an important distinction though, in the same way that `merge` handles nil but isn't called `merge-maybe`. Let me know if you feel strongly.

calherries replied:
I'm with Chris on this one, especially since this is now called `merge-base` not `inherit-base`. I find it intuitive that if a base doesn't exist this is a no-op, like `(merge x nil)`.

qnkhuat commented on src/metabase/models/setting.clj:
```diff
@@ -1079,6 +1086,14 @@
        (not (ns-in-test? (:namespace setting-definition)))))
 
+(defn- inherit-base
+  "Incorporate the values in `:base` as default values for `setting-options`"
```
```suggestion
  "Inherit the values in `:base` as default values for `setting-options` if exists"
```

I think we should be clearer about `base` is an optional thing

qnkhuat commented on src/metabase/models/setting.clj:
```diff
@@ -1208,7 +1229,7 @@
                                      (with-meta (meta setting-symbol)))
         setting-definition-symbol (gensym "setting-")]
-    `(let [setting-options#          (merge ~options ~setting-metadata)
+    `(let [setting-options#          (merge (#'inherit-base ~options) ~setting-metadata)
```
```suggestion
    `(let [setting-options#          (merge (inherit-base ~options) ~setting-metadata)
```

IMO, I'd rather make `inherit-base` public than using the var-quote like this.

crisptrutski replied:
I'm slightly torn, but will make the change. I don't feel that this belongs in the public API, but I also don't like exposing a "var" call in generated code, both aesthetically amd as it has performance overhead, although its mute for a form that only executes during compilation.

calherries replied:
How about making `merge-base` a local variable?

crisptrutski replied:
You know [how I feel about long functions](https://dev.to/kanani_nirav/the-five-lines-of-code-principle-why-less-is-more-in-programming-31j6) - you see I'm a recovering old school Ruby programmer. Having more code emitted in macro expansion also makes me uncomfortable about code bloat, but this is probably trauma from C++ and Rust where it can have a big impact on compile time and binary sizes. 

You may still have won me over that it's better than emitting code with `#'` though 😄 

I've made peace with just having it public though. Since merge is no longer recursive, exposing it gives users a way to do their own transitive extensions.

qnkhuat reviewed (approved):
Cool, same with Cal this is the approach I like :).
I have a couple of nits tho.

crisptrutski commented:
> Cool, same with Cal this is the approach I like :). I have a couple of nits tho.

I still haven't quite accepted it - a uuid is a type, not just a bag of options! 

Maybe next time I'll be more cautious about presenting options that aren't my favorite  - hoping future setting reopens this chestnut 😜

calherries commented:
> a uuid is a type, not just a bag of options!

Ha, maybe I'll come to agree with you after a few more years of changes to `defsetting`. I find it hard to see the point yet

calherries commented on test/metabase/models/setting_test.clj:
```diff
@@ -199,6 +199,31 @@
+   :default  "base-default"})
+
+(defsetting test-base-setting
+  "Setting to test the `:feature` property of settings. This only shows up in dev."
```
```suggestion
  "Setting to test the `:base` property of settings. This only shows up in dev."
```

calherries commented on test/metabase/models/setting_test.clj:
```diff
@@ -199,6 +199,31 @@
+      (testing "Uses non-overridden base options"
+        (is (= :csv (:type setting)))
+        (is (= :none (:setter setting)))
+        (is (nil? (resolve 'test-base-setting!)))
```
I think this test would be clearer and less brittle if we tested a single attribute with two settings, one with a top-level override and one without. Otherwise we're coupling the test to the behaviour of the setting properties more than we need to. e.g.

```
(def ^:private base-options
  {:default "base-default"})

(defsetting test-no-default-with-base-setting
  "Setting to test the `:base` property of settings. This only shows up in dev."
  :base base-options)

(defsetting test-default-with-base-setting
  "Setting to test the `:base` property of settings. This only shows up in dev."
  :visibility :internal
  :base base-options
  :default "top-level-default")

(deftest ^:parallel defsetting-with-base-test
  (testing "A setting which specifies some base options"
    (let [no-top-level-default-setting (setting/resolve-setting :test-no-default-with-base-setting)
          top-level-default-setting              (setting/resolve-setting :test-default-with-base-setting)]
      (testing "Uses base options when no top-level options are specified"
        (is (= "base-default" (:default no-top-level-default-setting)))
        (is (= "base-default" (no-top-level-default-setting))))
      (testing "Uses top-level options when specified"
        (is (= "top-level-default" (:default top-level-default-setting)))
        (is (= "top-level-default" (no-top-level-default-setting)))))))
```

calherries replied:
Also we don't need to use the messy kondo ignore if we can avoid using the setter

crisptrutski replied:
I tried to make the test as "real world" as possible, by closely resembling the uuid-nonce case, but without depending on the actual prod definition. I do really like your suggestion of reusing the same base between those settings, I agree it adds clarity.

I very intentionally do want to test the setter though, as the code needed to be reworked to avoid emitting a setter that just throws exceptions. This was in the parent PR but I still want some test coverage.

crisptrutski replied:
I broke out the setter to a separate test, as it doesn't need to test both settings, and it's nice to reduce the scope of the kondo ignore.
